pkg/registry/registry.go

```diff
@@ -346,3 +367,90 @@ func (reg *Registry) ModuleDownload() http.HandlerFunc {
+func (reg *Registry) ProviderVersions() http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		var (
+			namespace = chi.URLParam(r, "namespace")
+			name      = chi.URLParam(r, "name")
+		)
+
+		ver, err := reg.providerStore.ListProviderVersions(r.Context(), namespace, name)
+		if err != nil {
+			http.Error(w, http.StatusText(http.StatusNotFound), http.StatusNotFound)
+			reg.logger.Error("ListProviderVersions", zap.Errors("err", []error{err}))
+			return
+		}
+
+		err = json.NewEncoder(w).Encode(ver)
+		if err != nil {
+			http.Error(w, http.StatusText(http.StatusNotFound), http.StatusNotFound)
+			reg.logger.Error("ListProviderVersions", zap.Errors("err", []error{err}))
+			return
+		}
+
+		w.WriteHeader(http.StatusOK)
+	}
+}
+
+// ProviderDownload returns a handler that returns a download link for a specific version of a provider.
+// https://developer.hashicorp.com/terraform/internals/provider-registry-protocol#find-a-provider-package
+func (reg *Registry) ProviderDownload() http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		var (
+			namespace = chi.URLParam(r, "namespace")
+			name      = chi.URLParam(r, "name")
+			version   = chi.URLParam(r, "version")
+			os        = chi.URLParam(r, "os")
+			arch      = chi.URLParam(r, "arch")
+		)
+
+		ver, err := reg.providerStore.GetProviderVersion(r.Context(), namespace, name, version, os, arch)
+		if err != nil {
+			http.Error(w, http.StatusText(http.StatusNotFound), http.StatusNotFound)
+			reg.logger.Error("GetProviderVersion", zap.Errors("err", []error{err}))
+			return
+		}
+
+		err = json.NewEncoder(w).Encode(ver)
+		if err != nil {
+			http.Error(w, http.StatusText(http.StatusNotFound), http.StatusNotFound)
+			reg.logger.Error("GetProviderVersion", zap.Errors("err", []error{err}))
+			return
+		}
+
+		w.WriteHeader(http.StatusOK)
+	}
+}
+
+// ProviderAssetDownload returns a handler that returns a provider asset.
+// Used as a proxy when the provider is not in a public repository
+func (reg *Registry) ProviderAssetDownload() http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		var (
+			owner     = chi.URLParam(r, "namespace")
+			repo      = chi.URLParam(r, "name")
+			tag       = chi.URLParam(r, "version")
+			assetName = chi.URLParam(r, "assetName")
+		)
+
+		asset, err := reg.providerStore.GetProviderAsset(r.Context(), owner, repo, tag, assetName)
+		if err != nil {
+			http.Error(w, http.StatusText(http.StatusNotFound), http.StatusNotFound)
+			reg.logger.Error("GetProviderAsset", zap.Errors("err", []error{err}))
+			return
+		}
+
+		written, err := io.Copy(w, asset)
+		if err != nil {
+			http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
+			reg.logger.Error("GetProviderAsset", zap.Errors("err", []error{err}))
+			return
+		}
+		reg.logger.Debug(fmt.Sprintf("wrote %d bytes to response", written))
```

Comment:
can you add a message prefix here as well, so we know what function is spitting this out?